### PR TITLE
Exposing lucene 6.x minhash filter.

### LINKED ISSUE #20149: Expose the new min hashing / LSH filter
Lucene 6.2.0 added locality sensitive hashing; see  https://issues.apache.org/jira/browse/LUCENE-6968 for details.  We should expose this in Elasticsearch.

### PR DESCRIPTION
**Pull request by a2lin**
I've tried to expose the 6.x minhash filter, and wrote some documentation from my (very inexpert) understanding on how this works.

I was mostly playing around with it using the following scenario that I cribbed off of the lucene ticket.

result (hits)

```
{
  "hits":[
    {
      "_index":"lsh_test",
      "_type":"doc_type",
      "_id":"2",
      "_score":1.4709876,
      "_source":{
        "doc_lsh":"elasticsearch is also an open source enterprise search engine based on Lucene"
      }
    },
    {
      "_index":"lsh_test",
      "_type":"doc_type",
      "_id":"1",
      "_score":0.61897373,
      "_source":{
        "doc_lsh":"elasticsearch is also an open source search engine based on Lucene"
      }
    },
    {
      "_index":"lsh_test",
      "_type":"doc_type",
      "_id":"3",
      "_score":0.4758651,
      "_source":{
        "doc_lsh":"elasticsearch is also a popular open source enterprise search engine based on Lucene"
      }
    }
  ]
}
```

query:

```
{
  "match": "elasticsearch is also an open source search engine"
}
```

(bulk) indexed documents:

```
{ "index" : { "_index" : "lsh_test", "_type" : "doc_type", "_id" : "1" }}
{"doc_lsh": "elasticsearch is also an open source search engine based on Lucene"}
{ "index" : { "_index" : "lsh_test", "_type" : "doc_type", "_id" : "2" }}
{"doc_lsh": "elasticsearch is also an open source enterprise search engine based on Lucene"}
{ "index" : { "_index" : "lsh_test", "_type" : "doc_type", "_id" : "3" }}
{"doc_lsh": "elasticsearch is also a popular open source enterprise search engine based on Lucene"}
{ "index" : { "_index" : "lsh_test", "_type" : "doc_type", "_id" : "4" }}
{"doc_lsh": "Apache Lucene is a high-performance, full-featured text search engine library written entirely in Java"}
```

mapping:

```
{
  "settings":{
    "analysis":{
      "filter":{
        "five_shingle":{
          "type":"shingle",
          "max_shingle_size":5,
          "min_shingle_size":3,
          "output_unigrams":false
        },
        "minhash":{
          "type":"min_hash",
          "with_rotation":"false"
        }
      },
      "analyzer":{
        "lsh_analyzer":{
          "type":"custom",
          "tokenizer":"whitespace",
          "filter":[
            "five_shingle",
            "minhash"
          ]
        }
      }
    }
  },
  "mappings":{
    "doc_type":{
      "properties":{
        "doc_lsh":{
          "type":"text",
          "analyzer":"lsh_analyzer",
          "store":"true"
        }
      }
    }
  }
}
```

Closes #20149.

### CODE REVIEW

File: docs/reference/analysis/tokenfilters/minhash-tokenfilter.asciidoc [hunk 1]
```diff
@@ -0,0 +1,22 @@
+[[analysis-minhash-tokenfilter]]
+== Minhash Token Filter
+
+A token filter of type `minhash` hashes each token of the token stream and divides
```
Comment (jpountz) [inline docs/reference/analysis/tokenfilters/minhash-tokenfilter.asciidoc]: s/minhash/min_hash/

**Comment by clintongormley** [comment]
Hi @a2lin 

Thanks for the PR.  This needs documentation added too.  Btw, why are you using shingles of 3..5 words?  This will create many more tokens than you really need and isn't considered good practice.  Instead I'd just use shingles of length 2.

**Comment by a2lin** [comment]
@clintongormley Thanks for the advice. I poked around with 5 because the function comment for [MinHashFilter](https://git-wip-us.apache.org/repos/asf?p=lucene-solr.git;a=blob;f=lucene/analysis/common/src/java/org/apache/lucene/analysis/minhash/MinHashFilter.java;h=60df1c051c4730f4f4068628ced31704f1591866;hb=96a5595#l33) suggests that the expected incoming tokens are 5 word shingles.

Can you link me to an example of the documentation that this feature needs added? I could only find the shingled version of:

> docs/reference/analysis/tokenfilters/minhash-tokenfilter.asciidoc

 when I searched the code for the shingleTokenFilterFactory analogue.

**Comment by clintongormley** [comment]
@a2lin something like this https://www.elastic.co/guide/en/elasticsearch/reference/master/analysis-fingerprint-tokenfilter.html

**Comment by a2lin** [comment]
@clintongormley Oops, I thought that was generated from the file that @jpountz commented on. I'll look again.

**Comment by clintongormley** [comment]
Oh sorry @a2lin - I completely missed the asciidoc!

**Comment by s1monw** [comment]
ok to test

**Comment by s1monw** [comment]
@a2lin thanks for fixing this!!!!!

**Comment by a2lin** [comment]
@s1monw thanks for merging!

**Pull request merged**